File: token-validation-spring/src/main/java/no/nav/security/token/support/spring/validation/interceptor/BearerTokenClientHttpRequestInterceptor.java
package no.nav.security.token.support.spring.validation.interceptor;

import static no.nav.security.token.support.spring.validation.interceptor.TokenUtil.authorizationTokens;
import static org.springframework.http.HttpHeaders.AUTHORIZATION;

/*
 * THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
 * OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION
 * ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A
 * PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
 */
import java.io.IOException;

import org.springframework.http.HttpRequest;
import org.springframework.http.client.ClientHttpRequestExecution;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.http.client.ClientHttpResponse;

import no.nav.security.token.support.core.context.TokenValidationContextHolder;

public class BearerTokenClientHttpRequestInterceptor implements ClientHttpRequestInterceptor {

    private final TokenValidationContextHolder holder;

    public BearerTokenClientHttpRequestInterceptor(TokenValidationContextHolder holder) {
        this.holder = holder;
    }

    @Override
    public ClientHttpResponse intercept(HttpRequest request, byte[] body, ClientHttpRequestExecution execution)
            throws IOException {

        authorizationTokens(holder).ifPresent(t -> request.getHeaders().add(AUTHORIZATION, t));
        return execution.execute(request, body);
    }
}

File: token-validation-spring/src/main/kotlin/no/nav/security/token/support/spring/validation/interceptor/BearerTokenClientHttpRequestInterceptor.kt
package no.nav.security.token.support.spring.validation.interceptor

import java.io.IOException
import no.nav.security.token.support.core.JwtTokenConstants.AUTHORIZATION_HEADER
import no.nav.security.token.support.core.context.TokenValidationContextHolder
import org.slf4j.LoggerFactory
import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpRequestExecution
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.http.client.ClientHttpResponse

class BearerTokenClientHttpRequestInterceptor(private val holder: TokenValidationContextHolder) : ClientHttpRequestInterceptor {
    private val log = LoggerFactory.getLogger(BearerTokenClientHttpRequestInterceptor::class.java)

    @Throws(IOException::class)
    override fun intercept(req: HttpRequest, body: ByteArray, execution: ClientHttpRequestExecution): ClientHttpResponse {
        holder.getTokenValidationContext().apply {
            if (hasValidToken()) {
                log.debug("Adding tokens to Authorization header")
                req.headers.add(AUTHORIZATION_HEADER, issuers.joinToString { "${getJwtToken(it)?.asBearer()}" })
            }
            return execution.execute(req, body)
        }
    }
}